src/service/data_products/filtering.py

```diff
@@ -0,0 +1,167 @@
+            raise ValueError(f"Invalid type for range filter: {type_.value}")
+        self.type = type_
+        self.low = self._parse_val(low, "low")
+        self.high = self._parse_val(high, "high")
+        if self.low is None and self.high is None:
+            raise ValueError("At least one of the low or high values must be provided")
+        self.low_inclusive = low_inclusive
+        self.high_inclusive = high_inclusive
+        if self.low is not None and self.high is not None and (
+            self.low > self.high
+            or (self.low == self.high and (not self.low_inclusive or not self.high_inclusive))
+        ):
+            raise ValueError(f"The range {self.to_range_string()} excludes all values")
+
+    def to_range_string(self):
+        """
+        Create a range string from this instance. See `from_string` for a description of the
+        format.
+        """
+        ret = ""
+        if self.low is not None:
+            ret += "[" if self.low_inclusive else "("
+            ret += str(self.low)
+        ret += ","
+        if self.high is not None:
+            ret += str(self.high)
+            ret += "]" if self.high_inclusive else ")"
+        return ret
+
+    @classmethod
+    def _parse_inclusivity(self, part, start):
+        inclusive = False
+        if start:
+            if part.startswith("["):
+                inclusive = True
+            if part.startswith("(") or part.startswith("["):
+                part = part[1:]
+        else:
+            if part.endswith("]"):
+                inclusive = True
+            if part.endswith(")") or part.endswith("]"):
+                part = part[:-1]
+        return part, inclusive
+    
+    def _parse_val(self, val, name):
+        if val is None or val == "":
+            return None
+        if self.type == ColumnType.DATE:
+            try:
+                parser.isoparse(val)
+                return val
+            except ValueError as e:
+                raise ValueError(f"{name} value is not an ISO8601 date: {val}")
+        else:
+            try:
+                return float(val)
+            except ValueError as e:
+                raise ValueError(f"{name} value is not a number: {val}") from e
+
+    def __eq__(self, other: object) -> bool | NotImplementedType:
+        if not isinstance(other, RangeFilter):
+            return NotImplemented
+        return (self.type, self.low, self.high, self.low_inclusive, self.high_inclusive) == (
+            other.type, other.low, other.high, other.low_inclusive, other.high_inclusive)
+        
+    def __repr__(self):
+        l, h = self.low, self.high
+        if self.type == ColumnType.DATE:
+            l = f'"{l}"'
+            h = f'"{h}"'
+        return (f"RangeFilter(ColumnType.{self.type.name}, {l}, {h}, "
+            + f"{self.low_inclusive}, {self.high_inclusive})")
+
+    @classmethod
+    def from_string(cls, type_: ColumnType, string: str) -> Self:
```

Comment:
we didn't do any validation for `string`. Should we validate it before calling filter? i.e. If I pass `<2, 3)`, it will raise at `_parse_val` but we could raise earlier.

Reply:
got it. 👍